### PR TITLE
JAMES-2541 RabbitMQ mail queue

### PR DESCRIPTION
**Pull request by chibenwa**
Based on https://github.com/linagora/james-project/pull/1720

I did not tackle the subject of attribute serialization but the rest should be pretty fine...

### CODE REVIEW

File: core/pom.xml [hunk 1]
```diff
@@ -86,6 +86,10 @@
             <artifactId>junit-vintage-engine</artifactId>
```
Comment (chibenwa) [inline core/pom.xml]: This commit needs tests IMO...

File: core/src/test/java/org/apache/james/core/MailAddressTest.java [hunk 2]
```diff
@@ -1,241 +1,253 @@
...
+     * Test method for {@link MailAddress#MailAddress(javax.mail.internet.InternetAddress)}.
+     */
+    @Test
+    public void testMailAddressInternetAddress() {
+
+        try {
+            new MailAddress(new InternetAddress(GOOD_QUOTED_LOCAL_PART));
+        } catch (AddressException e) {
+            System.out.println("AddressException" + e.getMessage());
+            assertThat(false).describedAs(e.getMessage()).isTrue();
```
Comment (blackheaven) [inline core/src/test/java/org/apache/james/core/MailAddressTest.java]: `Assert.fail()` is more straightforward.

File: core/src/test/java/org/apache/james/core/MailAddressTest.java [hunk 3]
```diff
@@ -1,241 +1,253 @@
...
+     */
+    @Test
+    public void testGetDomain() {
+
+        try {
+            MailAddress a = new MailAddress(new InternetAddress(GOOD_ADDRESS));
+            assertThat(a.getDomain().equals(GOOD_DOMAIN)).describedAs(a.getDomain() + " != " + GOOD_DOMAIN).isTrue();
+        } catch (AddressException e) {
+            System.out.println("AddressException" + e.getMessage());
+            assertThat(false).describedAs(e.getMessage()).isTrue();
```
Comment (blackheaven) [inline core/src/test/java/org/apache/james/core/MailAddressTest.java]: `Assert.fail()` is more straightforward.

File: core/src/test/java/org/apache/james/core/MailAddressTest.java [hunk 4]
```diff
@@ -1,241 +1,253 @@
...
+     */
+    @Test
+    public void testGetLocalPart() {
+
+        try {
+            MailAddress a = new MailAddress(new InternetAddress(GOOD_QUOTED_LOCAL_PART));
+            assertThat(a.getLocalPart().equals(GOOD_LOCAL_PART)).describedAs(GOOD_LOCAL_PART + " != " + a.getLocalPart()).isTrue();
+        } catch (AddressException e) {
+            System.out.println("AddressException" + e.getMessage());
+            assertThat(false).describedAs(e.getMessage()).isTrue();
```
Comment (blackheaven) [inline core/src/test/java/org/apache/james/core/MailAddressTest.java]: `Assert.fail()` is more straightforward.

File: core/src/test/java/org/apache/james/core/MailAddressTest.java [hunk 5]
```diff
@@ -1,241 +1,253 @@
...
+     */
+    @Test
+    public void testToString() {
+
+        try {
+            MailAddress a = new MailAddress(new InternetAddress(GOOD_ADDRESS));
+            assertThat(a.toString().equals(GOOD_ADDRESS)).describedAs(a.toString() + " != " + GOOD_ADDRESS).isTrue();
+        } catch (AddressException e) {
+            System.out.println("AddressException" + e.getMessage());
+            assertThat(false).describedAs(e.getMessage()).isTrue();
```
Comment (blackheaven) [inline core/src/test/java/org/apache/james/core/MailAddressTest.java]: `Assert.fail()` is more straightforward.

File: core/src/test/java/org/apache/james/core/MailAddressTest.java [hunk 6]
```diff
@@ -1,241 +1,253 @@
...
+    public void testToInternetAddress() {
+
+        try {
+            InternetAddress b = new InternetAddress(GOOD_ADDRESS);
+            MailAddress a = new MailAddress(b);
+            assertThat(a.toInternetAddress().equals(b)).isTrue();
+            assertThat(a.toString().equals(GOOD_ADDRESS)).describedAs(a.toString() + " != " + GOOD_ADDRESS).isTrue();
+        } catch (AddressException e) {
+            System.out.println("AddressException" + e.getMessage());
+            assertThat(false).describedAs(e.getMessage()).isTrue();
```
Comment (blackheaven) [inline core/src/test/java/org/apache/james/core/MailAddressTest.java]: `Assert.fail()` is more straightforward.

File: server/blob/mail-store/src/test/java/org/apache/james/blob/mail/MimeMessageStoreTest.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            .setSubject("Important Mail")
+            .setText("Important mail content")
+            .build();
+
+        MimeMessagePartsId parts = testee.save(message).join();
+
+        MimeMessage retrievedMessage = testee.read(parts).join();
+
+        assertThat(MimeMessageUtil.asString(retrievedMessage))
+            .isEqualTo(MimeMessageUtil.asString(message));
```
Comment (blackheaven) [inline server/blob/mail-store/src/test/java/org/apache/james/blob/mail/MimeMessageStoreTest.java]: Why not compare directly `retrievedMessage` and `message`?

File: server/blob/mail-store/src/test/java/org/apache/james/blob/mail/MimeMessageStoreTest.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            .addToRecipient("toddy@any.com")
+            .setSubject("Important Mail")
+            .build();
+
+        MimeMessagePartsId parts = testee.save(message).join();
+
+        MimeMessage retrievedMessage = testee.read(parts).join();
+
+        assertThat(MimeMessageUtil.asString(retrievedMessage))
+            .isEqualTo(MimeMessageUtil.asString(message));
```
Comment (blackheaven) [inline server/blob/mail-store/src/test/java/org/apache/james/blob/mail/MimeMessageStoreTest.java]: idem

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/HeadersDto.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.apache.mailet.PerRecipientHeaders;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.Multimap;
+
+public class HeadersDto {
```
Comment (rouazana) [inline server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/HeadersDto.java]: I know this is long to write, but that's PerRecipientHeadersDTO, not HeadersDTO

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/MailDTOTest.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+                    .value("uedcgukrcg")
+                    .build(), MailAddressFixture.RECIPIENT2)
+                .state("state")
+                .name("mail-name-558")
+                .build(),
+            MimeMessagePartsId.builder()
+                .headerBlobId(BLOB_ID_FACTORY.from("210e7136-ede3-44eb-9495-3ed816d6e23b"))
+                .bodyBlobId(BLOB_ID_FACTORY.from("ef46c026-7819-4048-b562-3a37469191ed"))
+                .build());
+    }
```
Comment (rouazana) [inline server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/MailDTOTest.java]: maybe you can add a test with a minimalist mail (with all non mandatory fields empty)

File: core/src/test/java/org/apache/james/core/MailAddressTest.java [hunk 11]
```diff
@@ -1,241 +1,253 @@
...
+        assertThat(MailAddress.getMailSender(MailAddress.NULL_SENDER_AS_STRING))
+            .isEqualTo(MailAddress.nullSender());
+    }
+
+    @Test
+    public void getMailSenderShouldReturnParsedAddressWhenNotNullAddress() throws Exception {
+        assertThat(MailAddress.getMailSender(GOOD_ADDRESS))
+            .isEqualTo(new MailAddress(GOOD_ADDRESS));
+    }
+
```
Comment (rouazana) [inline core/src/test/java/org/apache/james/core/MailAddressTest.java]: please, try to commit white space changes in an other commit (maybe next time)

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/MailDTO.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+class MailDTO {
+
+    static MailDTO fromMail(Mail mail) {
+        return new MailDTO(
+            mail.getRecipients().stream()
+                .map(MailAddress::asString)
+                .collect(Guavate.toImmutableList()),
+            mail.getName(),
+            mail.getSender().asString());
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/MailDTO.java]: I choose to do it in a later commit for readability

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/MailDTO.java [hunk 13]
```diff
@@ -19,49 +19,148 @@
...
+        return remoteAddr;
+    }
+
+    @JsonProperty("remoteHost")
+    String getRemoteHost() {
+        return remoteHost;
+    }
+
+    @JsonProperty("perRecipientHeaders")
+    String getPerRecipientHeaders() {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/MailDTO.java]: Handled in a later commit...

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/HeadersDto.java (hunk 9 above)
Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/HeadersDto.java]: Nope this one is only a set of headers

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/MailDTO.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+import org.apache.james.core.MailAddress;
+import org.apache.mailet.Mail;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableList;
+
+class MailDTO {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/MailDTO.java]: IMO going beyond a simplistic public/private scopes strategy don't hurts, and might allow better interface readability (if it is not public hence it's not part of the API) so I'd support such a style.

File: server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+
+    @Disabled
+    @Override
+    public void queueShouldPreserveRemoteAddress() {
+
+    }
+
+    @Disabled
+    @Override
+    public void queueShouldPreserveRemoteHost() {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueTest.java]: Yes, in fine I believe I should have introduced the full DTO in a first commit then the MQ implem.

But at that time I had no tests for DTOs thus I choose another strategy...

File: core/src/test/java/org/apache/james/core/MailAddressTest.java (hunk 11 above)
Comment (chibenwa) [reply core/src/test/java/org/apache/james/core/MailAddressTest.java]: Yes maybe next time. Sorry, I did not even noticed it...

File: core/src/test/java/org/apache/james/core/MailAddressTest.java [hunk 16]
```diff
@@ -1,241 +1,253 @@
...
+    public void testMailAddressStringString() {
+
+        try {
+            new MailAddress("local-part", "domain");
+        } catch (AddressException e) {
+            assertThat(false).describedAs(e.getMessage()).isTrue();
+        }
+        try {
+            MailAddress a = new MailAddress("local-part", "-domain");
+            assertThat(true).describedAs(a.toString()).isFalse();
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply core/src/test/java/org/apache/james/core/MailAddressTest.java]: This change-set is due to white space changes... A not careful re-indent. Won't be fixed here. If you care bout this, we should fix it as part of https://github.com/linagora/james-project/pull/1721

File: server/blob/mail-store/src/test/java/org/apache/james/blob/mail/MimeMessageStoreTest.java (hunk 7 above)
Comment (chibenwa) [reply server/blob/mail-store/src/test/java/org/apache/james/blob/mail/MimeMessageStoreTest.java]: Java mail messed up equals and hashCode. Comparing the raw content for MimeMessages is a common practice in the code base

**Comment by chibenwa** [comment]
```
[c827f434206fdc24f1fda619805f506d94b7801c] [ERROR]   SerializationUtilTest.trySerializeShouldReturnString:61 expected:<"rO0ABXNyAE[1vcmcuYXBhY2hlLmphbWVzLnF1ZXVlLmptcy5KTVNTZXJpYWxpemF0aW9uVXRpbHNUZXN0JFNlcmlhbGl6YWJsZVN0cmluZ0hvbGRlcsy4/DEA8nRZ]AgABTAAFdmFsdWV0ABJM...> but was:<"rO0ABXNyAE[RvcmcuYXBhY2hlLmphbWVzLnV0aWwuU2VyaWFsaXphdGlvblV0aWxUZXN0JFNlcmlhbGl6YWJsZVN0cmluZ0hvbGRlcjk393aULV9X]AgABTAAFdmFsdWV0ABJM...>
```

crap

**Comment by chibenwa** [comment]
I would argue that class move = class serial id change = different serialiwation

Updating to the new value should fix it...

File: server/queue/queue-rabbitmq/src/test/resources/json/mail_min.json [hunk 17]
```diff
@@ -0,0 +1,16 @@
+{
+  "recipients":["recipient1@localhost","recipient2@localhost"],
+  "name":"mail-name-558",
+  "sender":"sender@localhost",
+  "state":"state",
+  "errorMessage":"an error",
```
Comment (rouazana) [inline server/queue/queue-rabbitmq/src/test/resources/json/mail_min.json]: is it really minimal?

File: server/queue/queue-rabbitmq/src/test/resources/json/mail_min.json [hunk 18]
```diff
@@ -1,16 +1,14 @@
...
-  "lastUpdated":1473344752.000000000,
-  "attributes":{"att1":"rO0ABXQABXZhbHVl"},
-  "remoteAddr":"159.221.12.145",
-  "remoteHost":"toto.com",
-  "perRecipientHeaders":{
-    "recipient1@localhost":{"header":{"X-custom-header":["uedcgukrcg"]}},
-    "recipient2@localhost":{"header":{"X-custom-header-2":["uedcgukrcg"]}}},
+  "sender":null,
+  "state":null,
+  "errorMessage":null,
```
Comment (rouazana) [inline server/queue/queue-rabbitmq/src/test/resources/json/mail_min.json]: oh, now it is :)

Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/test/resources/json/mail_min.json]: Sorry I messed up while I already pushed...

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**